econsa/correlation.py

```diff
@@ -0,0 +1,175 @@
+        all diagonal elements are 1,
+        and all eigenvalue is positive (default value is `True`).
+
+    force_calc : bool, optional
+        When `True`, calculate the covariances ignoring all special preset combinations of marginals
+        (default value is `False`).
+
+    Returns
+    -------
+    gc_corr : numpy.ndarray
+        The transformed correlation matrix that is ready to be fed into a Gaussian copula.
+
+    References
+    ----------
+    .. [K2012] Kucherenko, S., Tarantola, S., & Annoni, P. (2012).
+        Estimation of global sensitivity indices for models with
+        dependent variables. Computer Physics Communications, 183(4), 937–946.
+
+    .. [L1986] Liu, P.-L., & Der Kiureghian, A. (1986).
+        Multivariate distribution models with prescribed marginals
+        and covariances. Probabilistic Engineering Mechanics, 1(2), 105–112.
+
+    Examples
+    --------
+    >>> corr = [[1.0, 0.6, 0.3], [0.6, 1.0, 0.0], [0.3, 0.0, 1.0]]
+    >>> marginals = [cp.Normal(1.00), cp.Uniform(lower=-4.00), cp.Normal(4.20)]
+    >>> corr_transformed = gc_correlation(marginals, corr)
+    >>> copula = cp.Nataf(cp.J(*marginals), corr_transformed)
+    >>> corr_copula = np.corrcoef(copula.sample(1000000))
+    >>> np.testing.assert_almost_equal(corr, corr_copula, decimal=6)
+    """
+    corr = np.atleast_2d(corr)
+
+    # Test that marginals are all cp.distributions.
+    for marginal in marginals:
+        if issubclass(type(marginal), cp.distributions.baseclass.Dist):
+            continue
+        else:
+            raise NotImplementedError("marginals must be chaospy distributions")
+
+    if check_corr:
+        if not np.allclose(corr, corr.T):
+            raise ValueError("corr is not a symmetric matrix")
+        elif not np.all((corr >= -1) & (corr <= 1)):
+            raise ValueError("corr must be between 0 and 1")
+        elif not np.all(np.diagonal(corr) == 1):
+            raise ValueError("the diagonal of corr must all be 1")
+        elif np.all(np.linalg.eigvals(corr) > 0) == 0:
+            raise ValueError("corr is not positive-definite")
+
+    dim = len(corr)
+    indices = np.tril_indices(dim, -1)
+    gc_corr = np.identity(dim)
+
+    for i, j in list(zip(*indices)):
+        subset = [marginals[i], marginals[j]]
+        distributions, rho = cp.J(*subset), corr[i, j]
+        gc_corr[i, j] = _gc_correlation_pairwise(distributions, rho, force_calc)
+
+    # Align upper triangular with lower triangular.
+    gc_corr = gc_corr + gc_corr.T - np.diag(np.diag(gc_corr))
+
+    # If gc_corr is not positive definite, find the nearest one.
+    # The corr_nearest function is for finding positive semi-definite matrix,
+    # so sometimes it fails to find one. Loop until success.
+    if np.all(np.linalg.eigvals(gc_corr) > 0) == 0:
+        while True:
+            gc_corr_new = corr_nearest(gc_corr)
+            if np.all(np.linalg.eigvals(gc_corr_new) > 0) == 1:
```

Comment:
Do we really need the `while` loop? Isn't that always ensured by `corr_nearest`.

Reply:
Yes, since `corr_nearest` returns positive semi-definite matrix, sometimes it is exactly 0, which causes problem…